### DISCUSSION TITLE
exFAT support

### DISCUSSION THREAD

**Discussion by JORGETECH** [eduardofilo/RG350_adam_image#226] [Ideas & Requests]
Is there any clue on why exFAT formatted SD cards fail to mount? I know I can use FAT32 but exFAT is objectively better in every way and the kernel included with Adam should in theory have Samsung's exFAT mainlined drivers.

Does Adam image use udev to mount or something else? I want to try debugging the issue.

**Comment by neilswann80** [comment]
I'm almost certain this will be an issue with ODBeta rather than Adam Image specifically.  Even the ODBeta version of Commander fails to write to exFAT formatted partitions.

Have you tried on a generic ODBeta build?

**Comment by neilswann80** [comment]
@JORGETECH  I've confirmed with stock ODBeta, exFAT is not recognized by the system; so this is definitely not an issue specific to Adam Image.

 Did exFAT ever work on any of the handhelds Adam Image supports (RG-280V, RG-280M, RG-300X, RG-350, RG-350M, PocketGo2) even on the original OpenDingux?  I can only find reference to it being supported on the RetroMini (RS-90), but looking at the date even for this I think it would be pre-beta.

**Reply by JORGETECH** [reply]
exFAT was added to Ninoh-FOX's Rogue image, I believe it was never part of the stock OpenDingux system but it was an enhancement in the OpenRGH and Rogue forks. I suppose the code in those older forks is outdated by now.

**Comment by eduardofilo** [comment]
Actually ODb and therefore Adam support exFAT, but it seems that the udev rules are not suitable for mounting those drives properly. Doing manually does mount.

The relevant rule seems to be in `/etc/udev/rules.d/61-automount.rules`:

```
# Global mount options
ACTION=="add", ENV{mount_options}="noatime,nodiratime,nodev,nosuid"
# Filesystem specific options
ACTION=="add", PROGRAM=="/usr/sbin/blkid_fs %N", RESULT=="vfat", ENV{mount_options}="%E{mount_options},utf8,flush,uid=1000,gid=100"
ACTION=="add", PROGRAM=="/usr/sbin/blkid_fs %N", RESULT=="exfat", ENV{mount_options}="%E{mount_options}"
```

I will try to investigate to see if I can find the correct rule and how to override the previous one in the rootfs. If someone with more idea can find it first, please share it.

**Reply by eduardofilo** [reply]
The support does indeed appear to be installed, but the module is not loaded by default.:

```
rg280v:~ $ lsmod | grep exfat
rg280v:~ $ sudo modprobe exfat
rg280v:~ $ lsmod | grep exfat
exfat                  74701  0
rg280v:~ $ sudo cat /proc/filesystems | grep exfat
	exfat
```

After doing so, the card is mounted:

```
rg280v:~ $ mount | grep exfat
/dev/mmcblk1p1 on /media/sdcard type exfat (rw,nosuid,nodev,noatime,nodiratime,fmask=0022,dmask=0022,iocharset=utf8,errors=remount-ro)
```

**Reply by eduardofilo** [reply]
Creating the file `/media/data/local/etc/init.d/S01_loadexfat.sh` with the next content, it works, but apart from giving the script a more canonical form as discussed on [this site](https://unix.stackexchange.com/questions/396542/buildroot-how-to-load-modules-automatically) for example, I refuse to believe that there isn't a more correct way to load that module at startup. Someone who knows about buildroot would be needed here.

```
#!/bin/sh

modprobe exfat
/etc/init.d/S10udev stop
/etc/init.d/S10udev start
```

**Reply by JORGETECH** [reply]
Do you mean the udev rule works after mounting the exFAT module?

Regarding the loading of the module, udev may be able to load the module in a similar way as it's described on this ArchWiki article: [https://wiki.archlinux.org/title/Modalias#How_does_udev_work?](url)

EDIT: I think the other alternative is to have the exFAT driver built into the kernel instead of a module, which is possible by changing the kernel configuration in buildroot. My guess is that vfat support in included in such way.

**Reply by eduardofilo** [reply]
Yes, the rule works if the module is loaded. The udev rules cannot be changed without modifying the system rootfs, which I don't want to do. So I understand that there is no other way but to make a script as correct as possible to be executed from `/media/data/local/etc/init.d/`.

**Reply by eduardofilo** [reply]
> EDIT: I think the other alternative is to have the exFAT driver built into the kernel instead of a module, which is possible by changing the kernel configuration in buildroot. My guess is that vfat support in included in such way.

I don't want to do that either for the same reason. I promised the lead developer of ODbeta that we were not going to do a fork of his system.

**Reply by JORGETECH** [reply]
Since it seems that a clean solution would involve modifying stuff in the rootfs, I think the best thing to do right now is to either include the temporary "dirty" script in Adam or write information on how to fix the problem in the Readme (I consider the latter option to be good enough in this moment). In the meantime a discussion should be opened in the OpenDingux buildroot repo.

**Reply by eduardofilo** [reply]
I have created the FAQ number 12 in README to give more visibility to this topic.

https://github.com/eduardofilo/RG350_adam_image/#q12-is-exfat-format-supported-for-card-partitions

**Reply by JORGETECH** [reply]
Do you get permission issues when trying to create a folder or copy files using commander with exFAT filesystems? Commander complains every time I try to write to the external SD card formatted with exFAT which is really weird since the system was able to create the Adam folder structure on the external SD card.

EDIT: If I use mkdir with sudo on telnet I can create a directory, so I can confirm there's something really wrong with the permissions.

**Reply by JORGETECH** [reply]
EDIT: Ignore this last observation, the "uid" and "gid" is correctly set in the udev rules present in the rootfs, I cheked it with telnet. So it should do it right but something is failing along the way.

~~I checked again the automount rules you posted previously, I observed that the "uid" and "gid" options are not applied when mounting the exfat filesystem, this seems wrong to me since by default it should mount with the owner set to the root user, that would explain the permission issue.~~

~~Since it's something to be fixed in the udev rules I should report it to pcercuei, right?~~

**Reply by eduardofilo** [reply]
That's what I was going to say, that I seemed to remember that the partition was mounted with the owner and group indicated in those options, but I didn't actually try to create any files with the od user. I will check.

**Reply by eduardofilo** [reply]
@JORGETECH I have tested and had no problems creating files or directories with the od user. From Commander I have also been able to create a directory.

```
rg280v:/media $ id
uid=1000(od) gid=100(users) groups=5(tty),28(video),29(audio),100(users),1003(input),1005(kvm),1006(admin),1007(cpufreq),1008(mtp),1009(pm)
rg280v:/media $ mount | grep sdcard
/dev/mmcblk1p1 on /media/sdcard type exfat (rw,nosuid,nodev,noatime,nodiratime,uid=1000,gid=100,fmask=0022,dmask=0022,iocharset=utf8,errors=remount-ro)
rg280v:/media $ ls -l
drwxr-xr-x    4 root     root          4096 Feb 26  2022 data
drwxr-xr-x   10 od       users        32768 Jan  1  1980 sdcard
rg280v:/media $ cd sdcard/
rg280v:/media/sdcard $ ls -l
drwxr-xr-x    3 od       users        32768 Jan  1  1980 OpenBOR
drwxr-xr-x    2 od       users        32768 Jan  1  1980 apps
drwxr-xr-x    2 od       users        32768 Jan  1  1980 backups
drwxr-xr-x    3 od       users        32768 Jan  1  1980 bios
drwxr-xr-x    3 od       users        32768 Jan  1  1980 cheats
drwxr-xr-x   49 od       users        32768 Jan  1  1980 roms
rg280v:/media/sdcard $ touch nn
rg280v:/media/sdcard $ mkdir mm
rg280v:/media/sdcard $ ls -l
drwxr-xr-x    3 od       users        32768 Jan  1  1980 OpenBOR
drwxr-xr-x    2 od       users        32768 Jan  1  1980 apps
drwxr-xr-x    2 od       users        32768 Jan  1  1980 backups
drwxr-xr-x    3 od       users        32768 Jan  1  1980 bios
drwxr-xr-x    3 od       users        32768 Jan  1  1980 cheats
drwxr-xr-x    2 od       users        32768 Jan  1  1980 mm
-rwxr-xr-x    1 od       users            0 Jan  1  1980 nn
drwxr-xr-x   49 od       users        32768 Jan  1  1980 roms
```

**Reply by eduardofilo** [reply]
Copying files with Commander only gave me an error when copying a directory containing symbolic links, which makes sense. But copying another one that only has normal files gives no problems.
